src/algebra/algebra/non_unital.lean

```diff
@@ -0,0 +1,182 @@
+
+variables {R A} [non_unital_non_assoc_semiring A] [non_unital_non_assoc_algebra R A]
+
+lemma smul_mul_assoc (t : R) (a b : A) :
+  (t • a) * b = t • (a * b) :=
+non_unital_non_assoc_algebra.smul_mul_assoc' t a b
+
+lemma mul_smul_comm (t : R) (a b : A) :
+  a * (t • b) = t • (a * b) :=
+non_unital_non_assoc_algebra.mul_smul_comm' t a b
+
+/-- The algebra multiplication as a bilinear map. -/
+def lmul {R : Type u₁} {A : Type u₂}
+  [comm_semiring R] [non_unital_non_assoc_semiring A] [non_unital_non_assoc_algebra R A] :
+  A →ₗ[R] module.End R A :=
+linear_map.mk₂ R (*) add_mul smul_mul_assoc mul_add mul_smul_comm
+
+/-- If the underlying `non_unital_non_assoc_semiring` is actually a `semiring` we have an
+`algebra`. -/
+def to_algebra {R : Type u₁} {A : Type u₂}
+  [comm_semiring R] [semiring A] [non_unital_non_assoc_algebra R A] : algebra R A :=
+algebra.of_semimodule smul_mul_assoc mul_smul_comm
+
+end non_unital_non_assoc_algebra
+
+variables [non_unital_non_assoc_semiring A] [non_unital_non_assoc_algebra R A]
+variables [non_unital_non_assoc_semiring B] [non_unital_non_assoc_algebra R B]
+variables [non_unital_non_assoc_semiring C] [non_unital_non_assoc_algebra R C]
+
+set_option old_structure_cmd true
+
+/-- A morphism of non-associative algebras. -/
+structure non_unital_non_assoc_algebra_hom extends A →ₗ[R] B, mul_hom A B
+
+attribute [nolint doc_blame] non_unital_non_assoc_algebra_hom.to_linear_map
+attribute [nolint doc_blame] non_unital_non_assoc_algebra_hom.to_mul_hom
+
+namespace non_unital_non_assoc_algebra_hom
+
+variables {R A B C}
+
+instance {R : Type u₁} {A : Type u₂} {B : Type u₃}
+  [comm_semiring R] [ring A] [ring B] [algebra R A] [algebra R B] :
+  has_coe (A →ₐ[R] B) (non_unital_non_assoc_algebra_hom R A B) :=
+⟨λ f, { to_fun    := f,
+        map_smul' := f.map_smul,
+        ..f, }⟩
+
+instance : has_coe (non_unital_non_assoc_algebra_hom R A B) (A →ₗ[R] B) := ⟨to_linear_map⟩
+
+instance : has_coe (non_unital_non_assoc_algebra_hom R A B) (mul_hom A B) := ⟨to_mul_hom⟩
+
+lemma to_linear_map_eq_coe (f : non_unital_non_assoc_algebra_hom R A B) : f.to_linear_map = ↑f :=
+rfl
+
+lemma to_mul_hom_eq_coe (f : non_unital_non_assoc_algebra_hom R A B) : f.to_mul_hom = ↑f :=
+rfl
+
+/-- see Note [function coercion] -/
+instance : has_coe_to_fun (non_unital_non_assoc_algebra_hom R A B) := ⟨_, to_fun⟩
+
+initialize_simps_projections non_unital_non_assoc_algebra_hom (to_fun → apply)
+
+@[simp] lemma coe_mk (f : A → B) (h₁ h₂ h₃) :
+  ((⟨f, h₁, h₂, h₃⟩ : non_unital_non_assoc_algebra_hom R A B) : A → B) = f := rfl
+
+@[simp, norm_cast] lemma coe_to_linear_map (f : non_unital_non_assoc_algebra_hom R A B) :
+  ((f : A →ₗ[R] B) : A → B) = f :=
+rfl
+
+@[simp, norm_cast] lemma coe_to_mul_hom (f : non_unital_non_assoc_algebra_hom R A B) :
+  ((f : mul_hom A B) : A → B) = f :=
+rfl
+
+@[simp] lemma map_smul (f : non_unital_non_assoc_algebra_hom R A B) (c : R) (x : A) :
+  f (c • x) = c • f x :=
+linear_map.map_smul (f : A →ₗ[R] B) c x
```

Comment:
Untested, but shorter as:
```suggestion
f.to_linear_map.map_smul c x
```

Reply:
Is this definitely preferable? Maybe I'm being ridiculous used a coercion because it doesn't require me to be opinionated about _how_ to turn `f` into a linear map whereas if I invoke `.to_linear_map` I require that such a method exists. Happy to change it.

Reply:
Well, it's a proof so the actual definition you use doesn't really matter. For this specific case, `to_linear_map` is easier to type, shorter, and avoids a typeclass search.

Reply:
I get that it's a proof so who cares about the term but I'm imagining a hypothetical world in which `non_unital_non_assoc_algebra_hom` still has a coercion to a linear map but by a method of a different name.

In any case, I'll change it: definitely better to avoid typeclass search.